x/mongo/driver/topology/server.go

```diff
@@ -116,12 +116,15 @@ type Server struct {
 	// globalCtx should be created in NewServer and cancelled in Disconnect to signal that the server is shutting down.
 	// heartbeatCtx should be used for individual heartbeats and should be a child of globalCtx so that it will be
 	// cancelled automatically during shutdown.
-	heartbeatLock      sync.Mutex
-	conn               *connection
-	globalCtx          context.Context
-	globalCtxCancel    context.CancelFunc
-	heartbeatCtx       context.Context
-	heartbeatCtxCancel context.CancelFunc
+	heartbeatLock   sync.Mutex
+	conn            *connection
+	globalCtx       context.Context
+	globalCtxCancel context.CancelFunc
+
+	// Data required to cancel blocking operations while checking the status of
+	// the server during heartbeat monitoring.
+	cancelCheckSig chan struct{}
+	checkCanceled  atomic.Value
```

Comment:
Sometimes `checkCanceled` is guarded by `heartbeatLock` and sometimes it isn't, making it difficult to reason about when we need to hold a lock and when we don't. Does it make sense to simplify the access pattern to always require holding `heartbeatLock` to access `checkCanceled` (which would allow changing the type to `bool`, removing the type assertions as well)?

Reply:
I believe this has been resolved by the check-cancel logic refactor using the `contextListener`.